man/man8/pptsapd.8
.TH PPTSAPD 8
.\" @(#) $Header: /xtel/pp/pp-beta/man/man8/RCS/pptsapd.8,v 6.0 1991/12/18 20:44:03 jpo Rel $
.\"
.\" $Log: pptsapd.8,v $
.\" Revision 6.0  1991/12/18  20:44:03  jpo
.\" Release 6.0
.\"
.\"
.\"
.SH NAME
pptsapd \- server for PP channels.
.SH
.in +.5i
.ti -.5i
/usr/lib/pp/pptsapd
\%[-c\ address] \%[\-t\ tailor]
.in -.5i
(under rc.local)
.SH DESCRIPTION
.I Pptsapd
is a server that listens for requests from the \fIqmgr\fP to start up
particular channels. It is a PP specific version of the \fItsapd\fP(8)
in actuality. Once the channel is identified, the program part of the
channel is executed in standard ISODE style.
.PP
\fIPptsapd\fP responds to the following flag:
.TP
\-c address
Specifies a different presentation address to listen on. It defaults
to the value given in the PP tailor file.
.TP
\-t tailor
Use the given file as a tailor file rather than the default.
.SH FILES
The PP tailor file
.SH "SEE ALSO"
qmgr(8), tsapd(8)
.br
\fIThe PP Manual: Volume 1 \- Installation and Operation\fP



CHANGES









             Changes from PP version 5.2 to 6.0





_1.  _O_v_e_r_v_i_e_w

This  short  document  is  a   description   of   the   more
user/administrator  visible changes that have taken place in
PP between the last interim version (5.2) and the beta  ver-
sion  (6.0). This document is not exhaustive or complete but
is a guide to what to look out for. In general the following
may be assumed.

o+    The Queue files are compatible  between  5.2  and  6.0.
     Note  however that the Queue structure has changed. Old
     channels will have to be recompiled  to  recognise  the
     new format, but new channels will read old queue files.

o+    The qmgr protocol is compatible between 5.2 and 6.0

o+    The syntax of some of  the  table  files  has  changed.
     Where appropriate conversion scripts are provided.

o+    Some procedure calls have been  changed  or  have  dif-
     ferent parameters.

o+    the libdsap.a library is now required for compilation.

In summary, you should be careful when upgrading to the  6.0
release.   It  will  not  work  by  just  replacing  the 5.2
binaries with 6.0. For definite the domain table  will  HAVE
to be changed.

_2.  _D_e_t_a_i_l_e_d _c_h_a_n_g_e_s

This list is in no particular order of preference or  impor-
tance.

Domain Table
     The domain table syntax has been  changed  to  be  more
     explicit  and  has  been  brought into line with the OR
     table syntax. We feel this is clearer and more logical.
     Tools  that  process  the domain table have been taught
     the   new   format   and   there   are    scripts    in
     Tools/tables/compat  to convert between the old and new
     formats.  There is NO backwards compatibility here. Old
     format tables will just plain not work.

Aliases Table



                     December 18, 1991





                           - 2 -


     The aliases table syntax has changed to a  main  choice
     of  synonym or alias with further qualifiers specifying
     the address type of the synonym or alias.  The  changes
     are

          x:alias y         x:alias y
          x:synonym y       x:synonym y
          x:822-synonym y   x:synonym y 822
          x:x400-synonym y  x:synonym y x400
          x:822 y           x:alias y 822
          x:x400 y          x:alias y x400
     where any white space in  y  must  be  quoted  using  a
     backslash  There  is also a "external" qualifier.  This
     disables further lookups of synonyms  and  aliases  and
     can  be  used  for  a  main  site mail hub mapping from
     site_wide_name@localsite                             to
     local_machine_name@localmachine  and  vice  versa.  The
     use of external stops the looping alias  problem  which
     would otherwise occur.

Return of Contents
     Return of contents is now supported.  This  is  a  side
     effect  and  part  of  the  work on routing of delivery
     reports.

Use of pepsy
     Use of the PEPSY ASN.1 compiler rather  than  PEPY/POSY
     is  now  used  in  many  places. This makes for smaller
     binaries and much quicker compilation.

ISODE 7.0
     PP 6.0 is aligned to ISODE 7.0. It will not work  under
     ISODE 6.0.  If you are using the 6.0 release you should
     be using ISODE 7.0. It should however work under  ISODE
     6.8.

Tailor file
     There have been many changes to the  tailor  file.  The
     introduction of several new variables and the splitting
     up of header and body part types. These are

          dap_user     Directory username for DS lookups
          dap_passwd   Directory Passwd for same
          fsync        disable use of fsync(2)
          headertype   list of headers c.f. bodypart
          queuestruct  use hierarchical queue structure
          submit_addr  call submit over network
          diskuse      limit the disk space consumed by PP

     Channel tailoring also has a number of new variables

          drchan       replaces old outchan
          hdrin        headers inbound
          hdrout       headers outbound



                     December 18, 1991





                           - 3 -


          ininfo       inbound channel info
          inadr        inbound channel addr type
          insubadr     inbound channel sub address type
          intable      inbound channel table
          key          give alternative names to channel
          mtatable     list of known mtas for channel binding
          outinfo      outbound channel info
          outsubadr    outbound channel sub address
          outtable     outbound channel table
          solo-proc    channel requires address one at a time
          bad-sender-policyset the policy for unroutable senders
          check        set the checking mode of the channel

Channel pairing
     Channels may now be explicitly  paired  together.  This
     allows one definition to specify both an inbound and an
     outbound channel.  It also allows several logical chan-
     nels  for  one  inbound protocol by selecting a channel
     per MTA.

Timezones
     Many more lettered timezones are recognised. This  list
     is ambiguous at best and numeric forms should really be
     used except at user interfaces.

Delivery Reports
     Delivery reports are now routed back to their  origina-
     tor.  In 5.2 they were returned directly down the chan-
     nel they arrived on. This allows for one way  links  in
     general  makes  things  easier. It also allows protocol
     conversion so that, for instance, a delivery report  on
     an  X.400 message may go back via an RFC-822 if that is
     the only option.  A side effect of this is the  channel
     variable  outchan  has been renamed drchan and has dif-
     ferent semantics attached to it (see the manual).

Header parts
     Separation of headers and bodyparts for  each  channel.
     This  is  mainly a syntactic change but will allow more

Src/pptsapd/pptsapd.c
/* pptsapd.c: pp version of the tsap daemon */

# ifndef lint
static char Rcsid[] = "@(#)$Header: /xtel/pp/pp-beta/Src/pptsapd/RCS/pptsapd.c,v 6.0 1991/12/18 20:27:31 jpo Rel $";
# endif

/*
 * $Header: /xtel/pp/pp-beta/Src/pptsapd/RCS/pptsapd.c,v 6.0 1991/12/18 20:27:31 jpo Rel $
 *
 * $Log: pptsapd.c,v $
 * Revision 6.0  1991/12/18  20:27:31  jpo
 * Release 6.0
 *
 */



#include <errno.h>
#include <signal.h>
#include "util.h"
#include "chan.h"
#include <varargs.h>
#include <isode/manifest.h>
#include <sys/stat.h>
#include "sys.file.h"
#ifndef X_OK
#define X_OK    1
#endif

#include <isode/tpkt.h>

#ifdef  TCP
#include <isode/internet.h>
#endif
#ifdef  X25
#include <isode/x25.h>
#endif
#ifdef  CONS
#include <isode/cons.h>
#endif
#ifdef  TP4
#include <isode/tp4.h>
#endif
#include <isode/isoservent.h>
#include <isode/logger.h>
#include <isode/tailor.h>

/*  */

static int  debug = 1;
static int  nbits = FD_SETSIZE;

static char *myname = "pptsapd";

#define NTADDRS         FD_SETSIZE

static struct TSAPaddr *tz;


static void    adios (), advise ();

static char *getchanpgm ();
static void tsapd ();
static void ts_adios ();
static void ts_advise ();
static void arginit ();
static void envinit ();

extern int  errno;
extern char	*pptsapd_addr, *pptailor;
/*  */

/* ARGSUSED */

main (argc, argv, envp)
int     argc;
char  **argv,
      **envp;
{
    int	   vecp;
    char   *vec[4];
    register struct NSAPaddr  *na;
    struct TSAPdisconnect   tds;
    register struct TSAPdisconnect  *td = &tds;

    arginit (argc, argv);
    envinit ();

    for (na = tz -> ta_addrs; na < &tz->ta_addrs[tz->ta_naddr]; na++) {
	    switch (na -> na_type) {
		case NA_TCP:
		    advise (LLOG_NOTICE, NULLCP, "listening on TCP %s %d",
			    na2str (na), ntohs (na -> na_port));
		    break;

		case NA_X25:
		    advise (LLOG_NOTICE, NULLCP, "listening on X.25 %s %s",
			    na2str (na),
			    sel2str (na -> na_pid, na -> na_pidlen, 1));
		    break;

		case NA_BRG:
		    advise (LLOG_NOTICE, NULLCP,
			    "listening on X.25 (BRIDGE) %s %s", na2str (na),
			    sel2str (na -> na_pid, na -> na_pidlen, 1));
		    break;

		case NA_NSAP:
		    advise (LLOG_NOTICE, NULLCP, "listening on NS %s",
			    na2str (na));
		    break;

		default:
		    adios (NULLCP, "unknown network type 0x%x", na -> na_type);
		    /* NOTREACHED */
	    }
    }
    if (tz -> ta_selectlen)
	advise (LLOG_NOTICE, NULLCP, "   %s",
			sel2str (tz -> ta_selector, tz -> ta_selectlen, 1));
    if (TNetListen (tz, td) == NOTOK)
         ts_adios (td, LLOG_EXCEPTIONS, "listen failed");

    for (;;) {
	if (TNetAccept (&vecp, vec, 0, NULLFD, NULLFD, NULLFD, NOTOK, td)
		== NOTOK) {
	    ts_advise (td, LLOG_EXCEPTIONS, "accept failed");
	    continue;
	}

	if (vecp <= 0)
	    continue;

	switch (TNetFork (vecp, vec, td)) {
	    case OK:
		tsapd (vecp, vec);
		exit (1);
		/* NOTREACHED */

	    case NOTOK:
		ts_advise (td, LLOG_EXCEPTIONS, "fork failed");
	    default:
		break;
	}
    }
}

/*  */

static char buffer1[4096];
static char buffer2[32768];


static void  tsapd (vecp, vec)
int     vecp;
char  **vec;
{
    char    buffer[BUFSIZ];
    char    buf2[BUFSIZ];
    struct TSAPstart   tss;
    register struct TSAPstart *ts = &tss;
    struct TSAPdisconnect   tds;
    register struct TSAPdisconnect  *td = &tds;
    CHAN	*chan;
    char	*program;

/* begin UGLY */
    (void) strcpy (buffer1, vec[1]);
    (void) strcpy (buffer2, vec[2]);
/* end UGLY */

    if (TInit (vecp, vec, ts, td) == NOTOK) {
	ts_advise (td, LLOG_EXCEPTIONS, "T-CONNECT.INDICATION");
	return;
    }

    advise (LLOG_NOTICE, NULLCP,
	    "T-CONNECT.INDICATION: <%d, <%s, %s>, <%s, %s>, %d, %d>",
	    ts -> ts_sd,
	    na2str (ts -> ts_calling.ta_addrs),
	    sel2str (ts -> ts_calling.ta_selector,
		     ts -> ts_calling.ta_selectlen, 1),
	    na2str (ts -> ts_called.ta_addrs),
	    sel2str (ts -> ts_called.ta_selector,
		     ts -> ts_called.ta_selectlen, 1),
	    ts -> ts_expedited, ts -> ts_tsdusize);

     if (ts -> ts_called.ta_selectlen == 0) {
	(void) sprintf (buffer, "No TSEL present");
	goto out;
    }
    if ((chan = ch_nm2struct (ts -> ts_called.ta_selector)) == NULL) {
    	(void) sprintf (buffer, "PP Service %s not found",
		ts -> ts_called.ta_selector);
	goto out;
    }
    (void) strcpy (buf2, chan -> ch_progname);
    vecp = sstr2arg (buf2, 20, vec, " \t");
    if (vecp <= 0)
	    goto out;
    if ((program = getchanpgm (vec[0])) == NULLCP) {
    	(void) sprintf (buffer, "Can't locate program %s", vec[0]);
	goto out;
    }
    vec[vecp] = buffer1;
    vec[vecp+1] = buffer2;
    vec[vecp+2] = NULLCP;
    (void) execv (program, vec);
    (void) sprintf (buffer, "unable to exec %s (%d args): %s", program,
		 vecp + 2, sys_errname (errno));
    PP_OPER (NULLCP, ("%s", buffer));
out: ;
    advise (LLOG_EXCEPTIONS, NULLCP, "%s", buffer);
    if ((int)strlen (buffer) >= TD_SIZE)
	buffer[0] = NULL;
    (void) TDiscRequest (ts -> ts_sd, buffer, strlen (buffer) + 1, td);

    exit (1);
}

static char	*getchanpgm (pgm)
char	*pgm;
{
	char	buf[BUFSIZ];
	static char 	name[BUFSIZ];
	extern char *cmddfldir, *chndfldir;

	getfpath (cmddfldir, chndfldir, buf);
	getfpath (buf, pgm, name);

	return name;
}
/*  */

static void  ts_adios (td, code, event)
register struct TSAPdisconnect *td;
int	code;
char	*event;
{
	ts_advise (td, code, event);

	exit (1);
}

static void  ts_advise (td, code, event)
register struct TSAPdisconnect *td;
int     code;
char   *event;
{
    char    buffer[BUFSIZ];

    if (td -> td_cc > 0)
	(void) sprintf (buffer, "[%s] %*.*s",
		TErrString (td -> td_reason),
		td -> td_cc, td -> td_cc, td -> td_data);
    else
	(void) sprintf (buffer, "[%s]", TErrString (td -> td_reason));

    advise (code, NULLCP, "%s: %s", event, buffer);
}

/*  */

static void arginit (argc, argv)
int argc;
char **argv;
{
    register char  *ap;
    int opt;
    extern char *optarg;
    extern int optind;
    char *taddr = NULLCP;

    if (myname = rindex (*argv, '/'))
	myname++;
    if (myname == NULL || *myname == NULL)
	myname = *argv;

    while ((opt = getopt (argc, argv, "c:t:")) != EOF) {
	    switch (opt) {
		case 'c':
		    taddr = optarg;
		    break;
		case 't':
		    pptailor = optarg;
		    break;
		default:
		    adios (NULLCP, "Usage %s [-t tailor] [-c address]");
		    break;
	    }
    }
    sys_init (myname);

    if (taddr == NULLCP)
	    taddr = pptsapd_addr;
    if ((tz = str2taddr (taddr)) == NULLTA)
	    adios (NULLCP, "Unknown transport address '%s'", taddr);
}

static void envinit () {
    int     i,
	    sd;

    nbits = getdtablesize ();

    if (!(debug = isatty (2))) {
	for (i = 0; i < 5; i++) {
	    switch (fork ()) {
		case NOTOK: 
		    sleep (5);
		    continue;

		case OK: 
		    break;

		default: 
		    _exit (0);
	    }
	    break;
	}

	(void) chdir ("/");

	if ((sd = open ("/dev/null", O_RDWR)) == NOTOK)
	    adios ("/dev/null", "unable to read");
	if (sd != 0)
	    (void) dup2 (sd, 0), (void) close (sd);
	(void) dup2 (0, 1);
	(void) dup2 (0, 2);

#ifdef  TIOCNOTTY
	if ((sd = open ("/dev/tty", O_RDWR)) != NOTOK) {
	    (void) ioctl (sd, TIOCNOTTY, NULLCP);
	    (void) close (sd);
	}
#else
#ifdef  SYS5
	(void) setpgrp ();
	(void) signal (SIGINT, SIG_IGN);
	(void) signal (SIGQUIT, SIG_IGN);
#endif
#endif
    }
    else
	pp_log_norm -> ll_stat |= LLOGTTY;

#ifndef sun             /* damn YP... */
    for (sd = 3; sd < nbits; sd++)
	(void) close (sd);
#endif

    (void) signal (SIGPIPE, SIG_IGN);

    ll_hdinit (pp_log_norm, myname);
    advise (LLOG_NOTICE, NULLCP, "starting");
}

/*    ERRORS */

#ifndef lint
static void    adios (va_alist)
va_dcl
{
    va_list ap;

    va_start (ap);

    _ll_log (pp_log_norm, LLOG_FATAL, ap);

    va_end (ap);

    _exit (1);
}
#else
/* VARARGS */

static void    adios (what, fmt)
char   *what,
       *fmt;
{
    adios (what, fmt);
}
#endif


#ifndef lint
static void    advise (va_alist)
va_dcl
{
    int     code;
    va_list ap;

    va_start (ap);

    code = va_arg (ap, int);

    _ll_log (pp_log_norm, code, ap);

    va_end (ap);
}
#else
/* VARARGS */

static void    advise (code, what, fmt)
char   *what,
       *fmt;
int     code;
{
    advise (code, what, fmt);
}
#endif
